### DISCUSSION TITLE
[RFC] ConfigProvider.useConfig hooks

### DISCUSSION THREAD

**Discussion by zombieJ** [ant-design/ant-design#40025] [RFCs - archive]
# 📋 RFC

## Summary

ConfigProvider provides `useConfig` hooks for unique entry.

## Basic example

```diff
    const MyApp = () => {
--    const componentDisabled = useContext(ConfigProvider.DisabledContext);
--    const componentSize = useContext(ConfigProvider.SizeContext);
++    const { componentDisabled, componentSize } = ConfigProvider.useConfig();

      // ...
    };
```

## Motivation

There are too many Context export from ConfigProvider like `ConfigContext` or `SizeContext`, it's not well maintain if need more Context export.

ref #39648

## Detailed

To make this RFC done:
1. Deprecated origin `SizeContext` to warning user use `useConfig` instead.
2. Origin `ConfigContext` keep the same without warning.
3. New `useConfig` support only `componentDisabled` and ``componentSize` for first version.
4. Internal still use XXXContext for perf consider

## Drawbacks

No break of current usage. But get warning.

## Adoption strategy

1. Document update
2. feature in minor

**Comment by crazyair** [comment]
```tsx
// import { useConfig } from 'antd'

const { componentDisabled,  componentSize } = useConfig({ componentSize: 'small' })

```

**Reply by zombieJ** [reply]
default props with `{ componentSize: 'small' }` will always not correct with ConfigProvider config.

**Comment by xliez** [comment]
How to warning user use useConfig instead? Is this right way?

```typescript
/**
 * @deprecated Please use `ConfigProvider.useConfig().componentDisabled` instead only if in antd internal.
 */
const DisabledContext = React.createContext<DisabledType>(false);
```

**Reply by zombieJ** [reply]
Yes. TS `@deprecated` warning is necessary. Additionally, we should wrap Object.defProp to warning if developer get from ConfigProvider.

**Reply by xliez** [reply]
Thanks for your prompt.
Here is the pr https://github.com/ant-design/ant-design/pull/40215
Welcome your review 😀

**Reply by zombieJ** [reply]
Wow. So fast! I still need other collaborator confirm API naming standard : )

**Comment by zombieJ** [comment]
@ant-design/ant-design-collaborators Pls help to check `ConfigProvider.useConfig` name.

**Reply by kiner-tang** [reply]
useConfig +1

**Comment by RedJue** [comment]
I think you can add a `context` suffix, such as `useConfigContext`, so that users can intuitively know that the context is exported in the hook, which is more semantic.

**Comment by JarvisArt** [comment]
`useConfig` +1，全局化配置的值保存在 `context` 是 `ConfigProvider` 内部的逻辑，用户可以不用去关心。
`useConfig` 更抽象些，即使以后全局化配置的值不是保存在 `context` ，这个名称依然适用。
